Instant Search: Constrain tabbing to the Search overlay when visible.

a8ck3n opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #23396.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->

- Adds a second event listener that watches for tabs.
- Consolidates event listeners into helper functions.
- Fixes an issue where the overlay would eat any presses of the Escape key, even when not visible.
- Adds a hidden button at the end of the overlay as the "anchor". This is visible to screen readers and we use it as the test when deciding to wrap around. The button becomes visible when focused.

#### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Visit a site with Instant Search enabled.
2. Bring up the Overlay.
3. Tab through the controls until it wraps around.
4. Try shift-tabbing.
5. Confirm that tabbing is constrained to the overlay.

**Note:** Using the Enter key to dismiss the form will cause it to immediately reappear if the input field on the underlying page is active. Using the space key to dismiss works correctly. This is filed separately in #23394.

**Note:** There is some weird behaviour with the "Load more…" button which is not very user friendly at the moment. This should be addressed as part of #19035 and/or #23397.

Example of current behaviour:

https://user-images.githubusercontent.com/40267301/182823496-4bf894ab-baa5-42c6-add5-a67d30d50828.mp4

a8ck3n committed:
Constrain tabbing to the Search overlay when visible.

- Adds a second event listener that watches for tabs.
- Consolidates event listeners into helper functions.
- Fixes an issue where the overlay would eat any presses of the Escape key, even when not visible.

kangzj commented:
Hi @a8ck3n, awesome job! It works as described 👍 

Got a question tho, should we show the button at all? Could we just use the search input as the start of the trap?

<img width="512" alt="image" src="https://user-images.githubusercontent.com/1425433/182075075-bde2e8d5-ed4e-4f4c-b117-d363b74cc4d6.png">

Also I guess we could use some more explanation around the Tab trapping logic there 😄

a8ck3n commented:
Hey @kangzj, thanks for the feedback. Some As to your Qs below!

> Got a question tho, should we show the button at all? Could we just use the search input as the start of the trap?
> Also I guess we could use some more explanation around the Tab trapping logic there 😄

So by default, tabbing past the end of the overlay will highlight the next item in the DOM. Because the form is inserted at the bottom of the page, it should be the last item, which means we then tab out of the browser view and into the browser chrome (like the tool bar icons, address bar, inspector views, etc). So the idea here is to monitor the first and last elements in our overlay (the input field and the new Close Search button) and intervene when either of them gets focus.

We can't use the search input as the element we watch because it won't get focus again.

As an example, imagine the tab loop for a given page as having 30 items in it. Our input field is item 10 and our close button is item 18. Items 1 - 9 are links on the underlying page. Items 10 - 18 are in the overlay. Items 19 - 30 are browser chrome. So the logic here is to monitor any tab action involving item 10 (input field + shift-tab) or item 18 (close button + tab) and to refocus the loop as needed. In this way, we contain tabbing to items 10 - 18 as long as the overlay is visible.

Does that make the logic a bit clearer?

kangzj commented:
Hi @a8ck3n Thanks for the explanation 👍 

> We can't use the search input as the element we watch because it won't get focus again.

Gotcha. I understand we have to use an anchor as the end of the overlay so that we could start over from the search input again, but the approach changes the UI with the big close button.

I'm not sure this would work, just an idea here - could we consider using `tabindex` to make the `x` button the end of tabbing, so that we don't have to add another one?

kangzj commented:
@a8ck3n docs from [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex) tho, doesn't encourage that.

> Warning: Avoid using tabindex values greater than 0. Doing so makes it difficult for people who rely on assistive technology to navigate and operate page content. Instead, write the document with the elements in a logical sequence.

😓

a8ck3n commented:
Hey @kangzj. I'm just seeing that part of my explanation was messed up due to Markdown formatting the `~` and hiding some of the text. I've fixed my comment above but it should have read like this:

> As an example, imagine the tab loop for a given page as having 30 items in it. Our input field is item 10 and our close button is item 18. Items 1 - 9 are links on the underlying page. Items 10 - 18 are in the overlay. Items 19 - 30 are browser chrome. So the logic here is to monitor any tab action involving item 10 (input field + shift-tab) or item 18 (close button + tab) and to refocus the loop as needed. In this way, we contain tabbing to items 10 - 18 as long as the overlay is visible.

Regarding the `tabindex` idea:

> I'm not sure this would work, just an idea here - could we consider using `tabindex` to make the `x` button the end of tabbing, so that we don't have to add another one?

I actually tried a few ways of doing this but it doesn't result in a workable solution. For example, setting the `tabindex` to `1` does push the `x` button to the end of the tab loop but it doesn't come after the last element in the overlay as you might expect. Instead, it comes after the window chrome, meaning you have to tab through the toolbar, address bar, any open inspectors, etc, before getting to the `x` button. In that way, it feels very confusing and out of place.

Also, as you noted, it's considered against best practice.

> Gotcha. I understand we have to use an anchor as the end of the overlay so that we could start over from the search input again, but the approach changes the UI with the big close button.

This was the best approach I could come up with. I was thinking it might be okay because it is largely the same as the "skip link" recommendation used on a standard page. I was working from recommendations on this page:

https://webaim.org/techniques/skipnav/

I also feel that having a close button at the end of the overlay is actually quite convenient when tabbing through results, despite having two close buttons in the tab loop.

Another option that might work would be to use the "Search powered by Jetpack" as the anchor but we'd need to force that to always be visible and make it a proper button so it's in the tab loop by default in all browsers. With the default system settings, both Firefox and Safari ignore it.

The third option I considered was to just refocus on the input field anytime the target was outside the overlay containing DIV. This was less elegant though in that it didn't work with Shift-Tab (ie: tabbing in reverse). I could give this approach another look.

I'm definitely open to any other suggestions though. In my research, I mostly found recommendations for how overlays should behave but very few practical code suggestions. I think this approach satisfies the most common recommendations though.

a8ck3n committed:
Improve “focus-trap” logic.

Now accounts for clicks outside the overlay and refocuses on the first/last element of the overlay as appropriate.

a8ck3n committed:
Move the “focus-trap” anchor button…

Now inside the overlay and positioned at the bottom-right of the screen. Stays hidden until focused.

a8ck3n committed:
Use constants for shared class/id values.

These are used across a few components so better to use constants per the OVERLAY_CLASS_NAME example.

a8ck3n commented:
I've pushed up a few improvements and changed the button placement/styling. Would you mind having another look at this one @kangzj?

kangzj commented on projects/packages/search/src/instant-search/components/search-results.scss:
```diff
@@ -24,6 +24,28 @@ $modal-max-width-lg: 95%;
+		right: 0;
+		border: none;
+		background-color: inherit;
+		color: $color-text-light;
```
```suggestion
		color: inherit;
```
For dark theme, the text is not very visible, so probably just inherit here?

a8ck3n replied:
Nice catch. I didn't test this in Dark Mode so I overlooked the styling there. Will get this fixed up.

kangzj commented on projects/packages/search/src/instant-search/components/overlay.jsx:
```diff
@@ -14,6 +14,30 @@ const Overlay = props => {
 
+		const handleTabEvent = event => {
+			if ( event.key === 'Tab' ) {
+				const overlay = document.getElementsByClassName( OVERLAY_CLASS_NAME )[ 0 ];
+				const isInsideOverlay = overlay.contains( event.target );
+
+				// Looking up the searchInput assumes knowledge of another component.
+				// Not currently a constent as it's dynamically generated.
+				const searchInput = document.getElementById( 'jetpack-instant-search__box-input-1' );
+				const tabAnchor = document.getElementById( OVERLAY_FOCUS_ANCHOR_ID );
+				if ( event.shiftKey === true ) {
+					if ( event.target === searchInput || false === isInsideOverlay ) {
+						event.preventDefault();
+						tabAnchor.focus();
+					}
+				}
+				if ( event.shiftKey === false ) {
+					if ( event.target === tabAnchor || false === isInsideOverlay ) {
+						event.preventDefault();
+						searchInput.focus();
+					}
+				}
```
```suggestion
				const overlay = document.getElementsByClassName( OVERLAY_CLASS_NAME )[ 0 ];
				const isInsideOverlay = overlay.contains( event.target );

				if ( isInsideOverlay ) {
					// Looking up the searchInput assumes knowledge of another component.
					// Not currently a constant as it's dynamically generated.
					const searchInput = document.getElementById( 'jetpack-instant-search__box-input-1' );
					const tabAnchor = document.getElementById( OVERLAY_FOCUS_ANCHOR_ID );
					if ( event.shiftKey === true && event.target === searchInput ) {
						event.preventDefault();
						tabAnchor.focus();
					} else if ( event.shiftKey === false && event.target === tabAnchor ) {
						event.preventDefault();
						searchInput.focus();
					}
				}
```

kangzj replied:
##### Do you think the logic would be a bit clearer this way? 👆

##### Dynamic ID

> // Not currently a constant as it's dynamically generated.

I guess it might not be the best idea to rely on a dynamically generated value, because a refactoring would easily break this.

Possible solutions:

- Could we maybe change the dynamic ID to a constant one in the overlay (I don't see any particular reason why it's not)
- Or we could use a custom attribute to identify the input, like use a data attribute? We could then query the element with class + the attribute.

`<input data-focus-trap="start" class="jetpack-instant-search__box-input" ... />`

- Or the least we could do is to add a comment back in there, saying it's relied on for tab trapping.

Personally I would prefer the first choice 😄 Or you have any other idea?

##### Naming
Lastly, do you think it's better to name the variables like `startAnchorOfFocusTrapping` and `endAnchorOfFocusTrapping` (you probably have better ones) which would better explain why we are focusing on them on the conditions.

a8ck3n replied:
These suggested changes won't capture Tab events outside the overlay. Part of the logic is to make sure tabbing from the browser Chrome into the DOM is also captured. As an example, if you focus the address bar and then hit Tab, it will move to the first item on the page. The second Tab will be visible to us and we use that to refocus into the overlay.

> I guess it might not be the best idea to rely on a dynamically generated value, because a refactoring would easily break this.

Yes, I wasn't sure how best to address this. I'll do some testing with your suggestions. Option one sounds like the way to go.

> Lastly, do you think it's better to name the variables like […]

Yeah, I think you're right. I'll revisit this too.

kangzj replied:
> These suggested changes won't capture Tab events outside the overlay. 

Ah you're right. All good then 👍

kangzj replied:
Hmmm do we really want to trap focus even focus is outside the overlay? 🙋‍♂️

a8ck3n replied:
> Hmmm do we really want to trap focus even focus is outside the overlay? 🙋‍♂️

Yup! If not, the user has to tab though all the underlying obscured page links to get back into the overlay. The focus-trap is only in effect when the overlay is visible. From [the original issue](https://github.com/Automattic/jetpack/issues/23396):

> It is possible for a keyboard user to tab out of the search modal without closing it. Once this happens, there is no visual reference for the current location of focus, and it may be very difficult to return to the modal.

> While the modal is open, keyboard focus should be limited to only allow access to focusable elements that are inside the modal.

> See WCAG 2.4.3: https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-order.html

Those underlying tab targets should be inaccessible until the overlay is dismissed.

kangzj reviewed (changes requested):
Hi @a8ck3n I'm really happy about what it looks and works. Great efforts!

<img width="189" alt="image" src="https://user-images.githubusercontent.com/1425433/182972322-4f54a31f-20b4-4084-a718-69b3c2f386d4.png">

I'd be happy to approve if the button visibility in the dark theme is improved, which I provided a suggestion 😄 

Another thing I noticed is that, the focus is trapped in the overly as well in Customberg. I'm happy for it to be resolved in a follow up PR.

a8ck3n commented:
> Hi @a8ck3n I'm really happy about what it looks and works. Great efforts!

Thanks for the review and feedback, @kangzj! I feel like we're making some good progress on this. I'll push a few more changes to address your suggestions and create a secondary issue to disable the "trap" in Customburg.

More changes incoming!

a8ck3n committed:
Improve styling of close button in Dark theme.

a8ck3n committed:
Rename variables used in “focus trap”.

Just some additional clarity around why we are grabbing these specific elements.

a8ck3n committed:
Improve technique used to look up search input.

Use the class name (invariable) instead of the potentially changing ID and codify across components by using a const value.

a8ck3n commented:
New changes should address the following:

- Improved styling for the secondary Close button in the Dark theme.
- Better const names for the first and last elements in the "focus-trap" logic.
- Using an invariable class name (instead of dynamic ID) to look up the input field.
- Using a const value for said class name.

**Note:** Does not address this issue yet…

> Another thing I noticed is that, the focus is trapped in the overly as well in Customberg. I'm happy for it to be resolved in a follow up PR.

I'd like to address this in a follow-up PR. The accessibility of the Filter controls still have issues at mobile sizes. This is separate to the focus-trap though, so I think it should also be handled in a separate PR.

kangzj reviewed (approved):
It looks good to me 👍  So nice to see it completed, some great efforts  @a8ck3n
